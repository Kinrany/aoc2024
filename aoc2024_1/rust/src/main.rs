use std::io::{BufRead, BufReader};

use anyhow::Context;

fn input(input: impl BufRead) -> anyhow::Result<Vec<(i32, i32)>> {
    input
        .lines()
        .map(|result| {
            let s = result.context("failed to read line")?;
            let (a, b) = s.split_once("   ").context("invalid line")?;
            Ok((a.parse()?, b.parse()?))
        })
        .collect()
}

fn solution(input: Vec<(i32, i32)>) -> i32 {
    let (mut list1, mut list2): (Vec<_>, Vec<_>) = input.into_iter().unzip();
    list1.sort_unstable();
    list2.sort_unstable();
    list1
        .into_iter()
        .zip(list2)
        .map(|(a, b)| (a - b).abs())
        .sum()
}

fn main() -> anyhow::Result<()> {
    let answer = solution(input(BufReader::new(std::io::stdin()))?);
    println!("{answer}");
    Ok(())
}

#[test]
fn parse_input() {
    let txt = include_bytes!("../../example.txt");
    let input = input(txt.as_slice()).unwrap();
    insta::assert_debug_snapshot!(input);
}

#[test]
fn solve_example() {
    let input = vec![(3, 4), (4, 3), (2, 5), (1, 3), (3, 9), (3, 3)];
    assert_eq!(solution(input), 11);
}
