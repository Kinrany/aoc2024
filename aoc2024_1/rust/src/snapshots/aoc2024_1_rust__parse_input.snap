---
source: aoc2024_1/rust/src/main.rs
assertion_line: 37
expression: input
snapshot_kind: text
---
[
    (
        3,
        4,
    ),
    (
        4,
        3,
    ),
    (
        2,
        5,
    ),
    (
        1,
        3,
    ),
    (
        3,
        9,
    ),
    (
        3,
        3,
    ),
]
